src/send/mod.rs

```diff
@@ -0,0 +1,99 @@
+#[cfg(test)]
+mod test;
+
+use rocket::{
+  data::{self, FromData},
+  http::Status,
+  Data,
+  Outcome,
+  Request,
+};
+use rocket_contrib::{Json, Value};
+use std::{borrow::Cow, collections::HashMap};
+use validator::{self, ValidationError};
+
+#[derive(Debug, Deserialize)]
+struct Body
+{
+  text: String,
+  html: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+struct Email
+{
+  to: Vec<String>,
+  cc: Option<Vec<String>>,
+  subject: String,
+  body: Body,
+}
+
+impl FromData for Email
+{
+  type Error = ValidationError;
+
+  fn from_data(request: &Request, data: Data) -> data::Outcome<Self, Self::Error>
+  {
+    let result = Json::<Email>::from_data(request, data);
+    match result
+    {
+      Outcome::Success(json) =>
+      {
+        let email = json.into_inner();
+
+        if email.to.len() == 0
+        {
+          return fail();
+        }
+
+        if let Some(outcome) = validate_addresses(&email.to)
+        {
+          return outcome;
+        }
+
+        if let Some(ref cc) = email.cc
+        {
+          if let Some(outcome) = validate_addresses(&cc)
+          {
+            return outcome;
+          }
+        }
+
+        Outcome::Success(email)
```

Comment:
This whole block could use the function chaining a bit.
```rust
email.cc
    .map(|ref cc| { validate_addresses(&cc) })
    .unwrap_or(Outcome::Success(email))
```

Reply:
It's also possible Rust might not like my shorter version, since it might still be borrowing cc from email... the joy of lifetimes.